File: backend/app/schemas/activity.py
from datetime import datetime

from pydantic import BaseModel


class ActivityLogOut(BaseModel):
    id: int
    project_id: int | None
    actor_id: int
    actor_username: str
    action_type: str
    target_type: str
    target_id: int
    description: str
    created_at: datetime

    model_config = {"from_attributes": True}

File: frontend/src/api/activity.ts
import { apiFetch, parseErrorDetail } from "./client";

export interface ActivityLogEntry {
  id: number;
  project_id: number | null;
  actor_id: number;
  actor_username: string;
  action_type: string;
  target_type: string;
  target_id: number;
  description: string;
  created_at: string;
}

export async function listProjectActivity(projectId: number): Promise<ActivityLogEntry[]> {
  const res = await apiFetch(`/projects/${projectId}/activity`);
  if (!res.ok) throw new Error(await parseErrorDetail(res));
  return res.json();
}

export async function listGlobalActivity(): Promise<ActivityLogEntry[]> {
  const res = await apiFetch("/activity");
  if (!res.ok) throw new Error(await parseErrorDetail(res));
  return res.json();
}

export async function listDashboardActivity(): Promise<ActivityLogEntry[]> {
  const res = await apiFetch("/dashboard/activity");
  if (!res.ok) throw new Error(await parseErrorDetail(res));
  return res.json();
}
